### PR TITLE
subcase search support for case search

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
Support for doing subcase queries in Case Search and CLE via the 'XPath' queries.

There are two variations on the query which allow doing 'exists' queries as well as 'count' queries:
```
subcase-exists('parent', sex = 'm')
subcase-count('host', status='active') > 2
```

Spec: https://docs.google.com/document/d/1yqS-DqPRDHAf2Dy64J053vnu83vS2DY4eYTC1kTitwo/edit#

## Technical Summary
This includes some small refactoring of the 'xpath -> ES' code. In addition to adding support for the above expressions this PR also adds support for the `not` function: 7fd573ebac47b0bad82455cf23de076911083b0c

Best reviewed as a whole 🏠 

## Feature Flag
Case Search

### Safety story
The code changes are well covered by tests. These changes also only impact CLE and the '_xpath_query' case search filter.

### Automated test coverage
Added

### QA Plan
None

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/case_search/filter_dsl.py [hunk 1]
```diff
@@ -316,20 +324,69 @@ def _parse_normalize_subcase_query(node):
...
+        case_count -= 1
+
+    if count_op != "=":
+        count_op = ">"
+
+    index_indetifier = current_node.predicates[0].right
+    subcase_predicates = []
+    for node in current_node.predicates[1:]:
+        axis = node.left.axis if node.left.axis is not None else ""
+        subcase_predicates.append(f"{axis}{node.left.node_test}{node.op}'{node.right}'")
```
Comment (snopoke) [inline corehq/apps/case_search/filter_dsl.py:385]: You shouldn't need to process these at all, you can just return them as is:

```suggestion
    subcase_predicates = current_node.predicates[1:]
```

File: corehq/apps/case_search/filter_dsl.py [hunk 2]
```diff
@@ -316,20 +324,69 @@ def _parse_normalize_subcase_query(node):
...
+    #             str(right.left + right.op + right.right)
+    #         ]
+
+        # NOTES:
+        #  - instead of returning a tuple we could create a dataclass to make it easier to work with and
+        #    could encapsulate some functionality:
+        #       subcase_query.include_parent(subcase_count)
+        #       subcase_query.create_parent_filter(matching_parent_ids)
+
+    current_node = parse_xpath(node)
```
Comment (snopoke) [inline corehq/apps/case_search/filter_dsl.py:352]: I'd expect that the `node` passed in has already been parsed

**Review by snopoke — COMMENTED** [review]
Progress looks good. Generally need a bit more validation to ensure that we're getting what we expect so that we can fail gracefully instead of generating unexpected exceptions.

File: corehq/apps/case_search/filter_dsl.py (hunk 1 above)
Comment (dominicochoa-dimagi) [reply corehq/apps/case_search/filter_dsl.py:385]: So the test cases should be like:
```python3
(
            _check,
            "subcase_exists[identifier='parent'][@case_type='bob']",
            ("parent", [<BinaryExpression "@case_type='bob'">], ">", 0, False)
)
```

Comment (snopoke) [reply corehq/apps/case_search/filter_dsl.py]: yes, or you can keep it as a string and parse it in `_check` to convert to the AST

File: corehq/apps/case_search/xpath_functions/__init__.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+XPATH_VALUE_FUNCTIONS = {
+    'date': date,
+}
+
+
+XPATH_QUERY_FUNCTIONS = {
+    'not': not_,
+    'subcase-exists': subcase,
+    'subcase-count': subcase,
+}
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/__init__.py:15]: I'll need to spend a bit more time with this PR to wrap my head around it, but could you explain a little about the reasons for splitting apart these two types of functions?

Comment (snopoke) [reply corehq/apps/case_search/xpath_functions/__init__.py:15]: This is to distinguish between functions that return an ES filter and functions that mutate value:
```
# date is the only value function we support currently
dob > date('2022-01-01')

# but you can't use it on it's own, it has to be as part of another expression
date('2022-01-01')

# but you can use functions like not
not(a = 1)
```

File: corehq/apps/case_search/xpath_functions/subcase_functions.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        )
+
+    parent_case_id_counter = Counter()
+    for subcase in es_query.run().hits:
+        indices = [index for index in subcase['indices'] if index['identifier'] == subcase_query.index_identifier]
+        if indices:
+            parent_case_id_counter.update([indices[0]['referenced_id']])
+
+    if subcase_query.op == '>' and subcase_query.count <= 0:
+        return list(parent_case_id_counter)
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/subcase_functions.py:109]: nitpick: isn't this redundant with the line below?

File: corehq/apps/case_search/xpath_functions/subcase_functions.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        .source('indices')
+    )
+
+    if es_query.count() > MAX_RELATED_CASES:
+        raise TooManyRelatedCasesError(
+            _("The related case lookup you are trying to perform would return too many cases"),
+            serialize(subcase_query.subcase_filter)
+        )
+
+    parent_case_id_counter = Counter()
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/subcase_functions.py:102]: Did you look into using an ES aggregation for this?  I guess there are likely few subcases per parent on average, so I could see it not being worth it.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/case_search/xpath_functions/subcase_functions.py (hunk 4 above)
Comment (snopoke) [reply corehq/apps/case_search/xpath_functions/subcase_functions.py:109]: It is but it skips iterating through the data and checking the count for each one.

File: corehq/apps/case_search/xpath_functions/subcase_functions.py (hunk 5 above)
Comment (snopoke) [reply corehq/apps/case_search/xpath_functions/subcase_functions.py:102]: I did not consider that. I don't think there's a big advantage but I can test it out.

**Comment by snopoke** [comment]
@esoergel I've made a few changes:
* switched to using ES aggregations
* beefed up the test suite a bit
* allow using the subcase functions without additional filters: `subcase-exists('parent')` and `subcase-count('parent') > 1`

**Review by esoergel — APPROVED** [review]

**Pull request merged**